codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java

```diff
@@ -396,4 +404,226 @@ public TextHeaders clear() {
+        }
+
+        @Override
+        public TextHeaders set(CharSequence name, CharSequence... values) {
+            DefaultTextHeaders.super.set(name, commaSeparate(charSequenceEscaper, values));
+            return DefaultTextHeaders.this;
+        }
+
+        @Override
+        public TextHeaders set(CharSequence name, Iterable<? extends CharSequence> values) {
+            DefaultTextHeaders.super.set(name, commaSeparate(charSequenceEscaper, values));
+            return DefaultTextHeaders.this;
+        }
+
+        @Override
+        public TextHeaders setObject(CharSequence name, Object... values) {
+            DefaultTextHeaders.super.set(name, commaSeparate(objectEscaper, values));
+            return DefaultTextHeaders.this;
+        }
+
+        @Override
+        public TextHeaders setObject(CharSequence name, Iterable<?> values) {
+            DefaultTextHeaders.super.set(name, commaSeparate(objectEscaper, values));
+            return DefaultTextHeaders.this;
+        }
+
+        private TextHeaders addEscapedValue(CharSequence name, CharSequence escapedValue) {
+            CharSequence currentValue = DefaultTextHeaders.super.get(name);
+            if (currentValue == null) {
+                DefaultTextHeaders.super.add(name, escapedValue);
+            } else {
+                DefaultTextHeaders.super.set(name, commaSeparateEscapedValues(currentValue, escapedValue));
+            }
+            return DefaultTextHeaders.this;
+        }
+
+        private <T> CharSequence commaSeparate(ValueEscaper<T> escaper, T... values) {
+            StringBuilder sb = new StringBuilder(values.length * DEFAULT_VALUE_SIZE);
+            if (values.length > 0) {
+                int end = values.length - 1;
+                for (int i = 0; i < end; i++) {
+                    escaper.escape(sb, values[i]).append(", ");
+                }
+                escaper.escape(sb, values[end]);
+            }
+            return sb.toString();
+        }
+
+        private <T> CharSequence commaSeparate(ValueEscaper<T> escaper, Iterable<? extends T> values) {
+            StringBuilder sb = new StringBuilder();
+            Iterator<? extends T> iterator = values.iterator();
+            if (iterator.hasNext()) {
+                T next = iterator.next();
+                while (iterator.hasNext()) {
+                    escaper.escape(sb, next).append(", ");
+                    next = iterator.next();
+                }
+                escaper.escape(sb, next);
+            }
+            return sb.toString();
+        }
+
+        private CharSequence commaSeparateEscapedValues(CharSequence currentValue, CharSequence value) {
+            return new StringBuilder(currentValue.length() + 2 + value.length())
+                    .append(currentValue)
+                    .append(", ")
+                    .append(value).toString();
+        }
+
+        private StringBuilder quote(StringBuilder sb, CharSequence value) {
```

Comment:
I'm not sure that quoting and escaping are divisible operations.  Perhaps clarification of method names will help but I think this algorithm may some updates.  Here is a summary of what I think the two terms mean:

- Escape: Special characters need to be preceded with a "token" to delineate the following character is content, and not a control character.
- Quote: If a special character is found, and the value has not already been Escaped and Quoted, then the value needs to be wrapped in `"` characters.

See http://grepcode.com/file/repo1.maven.org/maven2/org.apache.commons/commons-lang3/3.1/org/apache/commons/lang3/StringEscapeUtils.java#StringEscapeUtils.escapeCsv%28java.lang.String%29 for a reference of what we likely need to do.  The comments on this method are also helpful.